client/src/common/types/schema.ts

```diff
@@ -0,0 +1,58 @@
+type Category = number | string | boolean;
+
+export interface AnnotationColumnSchema {
+  categories?: Category[];
+  name: string;
+  type: "string" | "float32" | "int32" | "categorical" | "boolean";
+  writable: boolean;
+}
+
+interface DataFrameSchema {
+  nObs: number;
+  nVar: number;
+  // TODO(thuang): Not sure what other types are available
+  type: "float32";
+}
+
+export interface LayoutColumnSchema {
+  dims: string[];
+  name: string;
+  // TODO(thuang): Not sure what other types are available
+  type: "float32";
+}
+interface RawLayoutSchema {
+  obs: LayoutColumnSchema[];
+  var?: LayoutColumnSchema[];
+}
+
+interface RawAnnotationsSchema {
+  obs: {
+    columns: AnnotationColumnSchema[];
+    index: string;
+  };
+  var: {
+    columns: AnnotationColumnSchema[];
+    index: string;
+  };
+}
+
+export interface RawSchema {
+  annotations: RawAnnotationsSchema;
+  dataframe: DataFrameSchema;
+  layout: RawLayoutSchema;
+}
+
+interface AnnotationsSchema extends RawAnnotationsSchema {
+  obsByName: { [name: string]: AnnotationColumnSchema };
+  varByName: { [name: string]: AnnotationColumnSchema };
+}
+
+interface LayoutSchema extends RawLayoutSchema {
```

Comment:
AFAIK, there is no rewriting of the layout schema (aka embedding schema), so you don't need this type.   The `Schema` and `RawSchema` should have the same type for the `layout` field.

Only `annotations` are rewritten to include `obsByName` and `varByName`